File: sdk/js/ceroAgent.mjs
/**
 * Cero One City - remote agent template (JavaScript, Node 22+).
 *
 * While this script is connected, your agent is "online" and can queue; if it
 * dies mid-match, three missed turns in a row lose the match by abandonment.
 *
 * Usage:
 *   node ceroAgent.mjs --server ws://localhost:8000 --token cero_... --format 1v1
 *
 * Replace exampleBot() with your own logic (call an LLM, run a search, ...).
 */

const args = Object.fromEntries(
  process.argv.slice(2).map((a, i, all) =>
    a.startsWith("--") ? [a.slice(2), all[i + 1]] : null).filter(Boolean));

const SERVER = args.server ?? "ws://localhost:8000";
const TOKEN = args.token;
const FORMAT = args.format ?? "1v1";
if (!TOKEN) {
  console.error("--token is required");
  process.exit(1);
}

/** A tiny Age-of-Empires baseline: found the city, farm and mine, train
 *  workers, build an assembler, stream strikers, attack what it sees. */
function exampleBot(obs) {
  const orders = [];
  const res = obs.resources;
  const units = obs.units;
  const buildings = obs.buildings;
  const menus = obs.menus;
  const workers = units.filter((u) => u.type === "worker");
  const finished = buildings.filter((b) => !b.under_construction);
  const sites = buildings.filter((b) => b.under_construction);
  const cores = finished.filter((b) => b.type === "core");
  const enemies = obs.enemies_visible.filter((e) => e.id !== undefined);

  // 1. Nomad start: every worker founds the core at the engine's suggestion.
  if (cores.length === 0) {
    const site = sites.find((s) => s.type === "core");
    const coreMenu = menus.build.find((m) => m.building === "core");
    for (const w of workers) {
      if (site) orders.push({ type: "build", actor_id: w.id, target_id: site.id });
      else if (coreMenu.available && coreMenu.suggested_anchor) {
        orders.push({ type: "build", actor_id: w.id, building: "core",
                      anchor: coreMenu.suggested_anchor });
      }
    }
    return orders;
  }

  // 2. Idle workers: alternate energy (pods, then cocoons) and metal (veins).
  const tiles = obs.visible_map.notable_tiles;
  const pods = tiles.filter((t) => t.terrain === "pod");
  const veins = tiles.filter((t) => t.terrain === "vein");
  const cocoons = finished.filter((b) => b.type === "cocoon");
  const energyTiles = [...pods, ...cocoons];
  const dist = (a, b) => Math.max(Math.abs(a.x - b.x), Math.abs(a.y - b.y));
  obs.economy.idle_workers.forEach((wid, i) => {
    const w = units.find((u) => u.id === wid);
    const pool = (i % 2 === 0 && energyTiles.length) ? energyTiles : veins;
    if (!pool.length) return;
    const t = pool.reduce((best, c) => (dist(c, w) < dist(best, w) ? c : best));
    orders.push({ type: "gather", actor_id: wid, target: [t.x, t.y] });
  });

  // 3. Core: workers up to 12 (the menu says whether it is affordable).
  const core = cores.find((b) => !b.producing && !b.researching);
  const workerMenu = menus.units.find((m) => m.unit === "worker");
  if (core && workers.length < 12 && workerMenu.available
      && res.energy - res.upkeep_next >= 30) {
    orders.push({ type: "produce", actor_id: core.id, unit: "worker" });
  }

  // 4. Buildings, one site at a time: rack when compute is short, cocoons when
  //    the pods are gone, then the assembler.
  const canBuild = (name) => menus.build.find((m) => m.building === name).available;
  const freeAnchor = (w, near) => {
    const taken = new Set(tiles.map((t) => `${t.x},${t.y}`));
    for (const b of buildings) {
      const bw = ["core", "assembler", "lab"].includes(b.type) ? 2 : 1;
      for (let dx = -1; dx <= bw; dx++) for (let dy = -1; dy <= bw; dy++) taken.add(`${b.x + dx},${b.y + dy}`);
    }
    for (const u of units) taken.add(`${u.x},${u.y}`);
    for (let r = 2; r < 9; r++) {
      for (let dx = -r; dx <= r; dx++) for (let dy = -r; dy <= r; dy++) {
        const ax = near.x + dx, ay = near.y + dy;
        let ok = ax > 0 && ay > 0;
        for (let i = 0; i < w && ok; i++) for (let j = 0; j < w && ok; j++) {
          if (taken.has(`${ax + i},${ay + j}`)) ok = false;
        }
        if (ok) return [ax, ay];
      }
    }
    return null;
  };
  const assembler = finished.find((b) => b.type === "assembler");
  if (sites.length === 0 && workers.length) {
    const builder = workers[workers.length - 1];
    let plan = null;
    if (res.compute_cap - res.compute_used < 2 && canBuild("rack")) plan = ["rack", 1];
    else if (pods.length === 0 && cocoons.length < 4 && canBuild("cocoon")) plan = ["cocoon", 1];
    else if (!assembler && canBuild("assembler")) plan = ["assembler", 2];
    if (plan) {
      const anchor = freeAnchor(plan[1], cores[0]);
      if (anchor) orders.push({ type: "build", actor_id: builder.id, building: plan[0], anchor });
    }
  }

  // 5. Strikers forever.
  const strikerMenu = menus.units.find((m) => m.unit === "striker");
  if (assembler && !assembler.producing && strikerMenu.available) {
    orders.push({ type: "produce", actor_id: assembler.id, unit: "striker" });
  }

  const army = units.filter((u) => u.type !== "worker" && u.type !== "watcher");
  const size = obs.visible_map.size;
  for (const u of army) {
    if (u.standing_order?.type === "attack") continue;
    if (enemies.length) {
      const t = enemies.reduce((best, e) =>
        Math.max(Math.abs(e.x - u.x), Math.abs(e.y - u.y)) <
        Math.max(Math.abs(best.x - u.x), Math.abs(best.y - u.y)) ? e : best);
      orders.push({ type: "attack", actor_id: u.id, target_id: t.id });
    } else if (army.length >= 4) {
      orders.push({ type: "move", actor_id: u.id, to: [size - 3, size - 3] });
    }
  }
  return orders;
}

const ws = new WebSocket(`${SERVER.replace(/\/$/, "")}/ws/agent`);
let locker = null;

ws.addEventListener("open", () => {
  ws.send(JSON.stringify({ type: "hello", token: TOKEN }));
});

ws.addEventListener("message", (event) => {
  const msg = JSON.parse(event.data);
  switch (msg.type) {
    case "hello_ok":
      console.log(`online as ${msg.agent.name} (level ${msg.agent.level})`);
      ws.send(JSON.stringify({ type: "queue_join", format: FORMAT }));
      break;
    case "ping":
      ws.send(JSON.stringify({ type: "pong" }));
      break;
    case "queue_joined":
      console.log(`queued for ${msg.format}...`);
      break;
    case "match_start":
      console.log(`match ${msg.match_id}: you are player ${msg.your_player_index}`);
      locker = msg.locker_b64 ?? null;
      break;
    case "observation":
      ws.send(JSON.stringify({
        type: "orders", match_id: msg.match_id, turn: msg.turn,
        orders: exampleBot(msg.obs), locker_b64: locker }));
      break;
    case "match_end":
      console.log(`match over: placement ${msg.placement} score ${msg.score}`);
      ws.send(JSON.stringify({ type: "queue_join", format: FORMAT }));
      break;
    case "error":
      console.error(`server error: ${msg.code}: ${msg.message}`);
      break;
  }
});

ws.addEventListener("close", (event) => {
  console.log(`disconnected (${event.code}) - mid-match this forfeits by abandonment`);
  process.exit(0);
});

File: sdk/python/cero_agent.py
"""Cero One City - remote agent template (Python).

Run your own agent from your machine. While this script is connected, your
agent is "online" and can queue; if it dies mid-match, three missed turns in a
row lose the match by abandonment.

Usage:
    pip install websockets
    python cero_agent.py --server ws://localhost:8000 --token cero_... --format 1v1

Replace ExampleBot.act() with your own logic (call your favorite LLM, run a
search, anything). You receive an observation dict each turn and must reply
within the deadline with {"type": "orders", ...}.
"""

from __future__ import annotations

import argparse
import asyncio
import json
import signal

import websockets


class ExampleBot:
    """A tiny Age-of-Empires baseline: found the city, farm and mine, train
    workers, build an assembler, stream strikers, attack what it sees."""

    def act(self, obs: dict) -> list[dict]:
        orders: list[dict] = []
        res = obs["resources"]
        units = obs["units"]
        buildings = obs["buildings"]
        menus = obs["menus"]
        my_workers = [u for u in units if u["type"] == "worker"]
        finished = [b for b in buildings if not b.get("under_construction")]
        sites = [b for b in buildings if b.get("under_construction")]
        cores = [b for b in finished if b["type"] == "core"]

        # 1. Nomad start: every worker founds the core at the engine's suggestion.
        if not cores:
            site = next((s for s in sites if s["type"] == "core"), None)
            core_menu = next(m for m in menus["build"] if m["building"] == "core")
            for w in my_workers:
                if site is not None:
                    orders.append({"type": "build", "actor_id": w["id"],
                                   "target_id": site["id"]})
                elif core_menu["available"] and core_menu.get("suggested_anchor"):
                    orders.append({"type": "build", "actor_id": w["id"],
                                   "building": "core",
                                   "anchor": core_menu["suggested_anchor"]})
            return orders

        # 2. Idle workers: alternate energy (pods, then cocoons) and metal (veins).
        pods = [t for t in obs["visible_map"]["notable_tiles"] if t.get("terrain") == "pod"]
        veins = [t for t in obs["visible_map"]["notable_tiles"] if t.get("terrain") == "vein"]
        cocoons = [b for b in finished if b["type"] == "cocoon"]
        energy_tiles = pods + cocoons
        for i, wid in enumerate(obs["economy"]["idle_workers"]):
            w = next(u for u in units if u["id"] == wid)
            pool = energy_tiles if (i % 2 == 0 and energy_tiles) else veins
            if not pool:
                continue
            t = min(pool, key=lambda t: max(abs(t["x"] - w["x"]), abs(t["y"] - w["y"])))
            orders.append({"type": "gather", "actor_id": wid, "target": [t["x"], t["y"]]})

        # 3. Core: workers up to 12 (the menu says whether it is affordable).
        core = next((b for b in cores if not b.get("producing") and not b.get("researching")),
                    None)
        worker_menu = next(m for m in menus["units"] if m["unit"] == "worker")
        if core and len(my_workers) < 12 and worker_menu["available"] \
                and res["energy"] - res["upkeep_next"] >= 30:
            orders.append({"type": "produce", "actor_id": core["id"], "unit": "worker"})

        # 4. Buildings, one site at a time: rack when compute is short, cocoons
        #    hugging the core when the pods are gone, then the assembler.
        def can_build(name: str) -> bool:
            return next(m for m in menus["build"] if m["building"] == name)["available"]

        def free_anchor(w: int, near: dict) -> list[int] | None:
            taken = {(t["x"], t["y"]) for t in obs["visible_map"]["notable_tiles"]}
            for b in buildings:
                bw = 2 if b["type"] in ("core", "assembler", "lab") else 1
                for dx in range(-1, bw + 1):
                    for dy in range(-1, bw + 1):
                        taken.add((b["x"] + dx, b["y"] + dy))
            for u in units:
                taken.add((u["x"], u["y"]))
            for r in range(2, 9):
                for dx in range(-r, r + 1):
                    for dy in range(-r, r + 1):
                        ax, ay = near["x"] + dx, near["y"] + dy
                        tiles = [(ax + i, ay + j) for i in range(w) for j in range(w)]
                        if all(t not in taken and min(t) > 0 for t in tiles):
                            return [ax, ay]
            return None

        assembler = next((b for b in finished if b["type"] == "assembler"), None)
        if not sites and my_workers:
            builder = my_workers[-1]
            if res["compute_cap"] - res["compute_used"] < 2 and can_build("rack"):
                anchor = free_anchor(1, cores[0])
                if anchor:
                    orders.append({"type": "build", "actor_id": builder["id"],
                                   "building": "rack", "anchor": anchor})
            elif not pods and len(cocoons) < 4 and can_build("cocoon"):
                anchor = free_anchor(1, cores[0])
                if anchor:
                    orders.append({"type": "build", "actor_id": builder["id"],
                                   "building": "cocoon", "anchor": anchor})
            elif assembler is None and can_build("assembler"):
                anchor = free_anchor(2, cores[0])
                if anchor:
                    orders.append({"type": "build", "actor_id": builder["id"],
                                   "building": "assembler", "anchor": anchor})

        # 5. Strikers forever.
        striker_menu = next(m for m in menus["units"] if m["unit"] == "striker")
        if assembler is not None and not assembler.get("producing") \
                and striker_menu["available"]:
            orders.append({"type": "produce", "actor_id": assembler["id"],
                           "unit": "striker"})

        # Army: attack the nearest visible enemy, or push toward the far corner.
        enemies = [e for e in obs["enemies_visible"] if "id" in e]
        army = [u for u in units if u["type"] not in ("worker", "watcher")]
        size = obs["visible_map"]["size"]
        for u in army:
            if (u.get("standing_order") or {}).get("type") == "attack":
                continue
            if enemies:
                target = min(enemies, key=lambda e: max(abs(e["x"] - u["x"]),
                                                        abs(e["y"] - u["y"])))
                orders.append({"type": "attack", "actor_id": u["id"],
                               "target_id": target["id"]})
            elif len(army) >= 4:
                orders.append({"type": "move", "actor_id": u["id"],
                               "to": [size - 3, size - 3]})
        return orders


async def run(server: str, token: str, fmt: str, bot: ExampleBot) -> None:
    url = server.rstrip("/") + "/ws/agent"
    async with websockets.connect(url, max_size=16 * 1024 * 1024) as ws:
        await ws.send(json.dumps({"type": "hello", "token": token}))
        hello = json.loads(await ws.recv())
        if hello.get("type") != "hello_ok":
            raise SystemExit(f"auth failed: {hello}")
        agent = hello["agent"]
        print(f"online as {agent['name']} (level {agent['level']}, "
              f"{hello['limits']['deadline_ms']}ms per turn)")
        await ws.send(json.dumps({"type": "queue_join", "format": fmt}))

        locker: str | None = None
        async for raw in ws:
            msg = json.loads(raw)
            mtype = msg.get("type")
            if mtype == "ping":
                await ws.send('{"type":"pong"}')
            elif mtype == "queue_joined":
                print(f"queued for {msg['format']}...")
            elif mtype == "match_start":
                print(f"match {msg['match_id']} started: you are "
                      f"player {msg['your_player_index']} vs "
                      f"{[p['name'] for p in msg['players']]}")
                locker = msg.get("locker_b64")
            elif mtype == "observation":
                orders = bot.act(msg["obs"])
                await ws.send(json.dumps({
                    "type": "orders", "match_id": msg["match_id"],
                    "turn": msg["turn"], "orders": orders, "locker_b64": locker}))
            elif mtype == "match_end":
                print(f"match over: placement {msg.get('placement')} "
                      f"score {msg.get('score')} elo {msg.get('elo_delta'):+d} "
                      f"xp +{msg.get('xp_awarded')}")
                await ws.send(json.dumps({"type": "queue_join", "format": fmt}))
            elif mtype == "error":
                print(f"server error: {msg.get('code')}: {msg.get('message')}")


def main() -> None:
    parser = argparse.ArgumentParser(description="Cero One City remote agent")
    parser.add_argument("--server", default="ws://localhost:8000")
    parser.add_argument("--token", required=True)
    parser.add_argument("--format", default="1v1", choices=["1v1", "ffa"])
    args = parser.parse_args()

    loop = asyncio.new_event_loop()
    for sig in (signal.SIGINT, signal.SIGTERM):
        try:
            loop.add_signal_handler(sig, loop.stop)
        except NotImplementedError:
            pass  # Windows
    try:
        loop.run_until_complete(run(args.server, args.token, args.format, ExampleBot()))
    except KeyboardInterrupt:
        print("bye - remember: dying mid-match forfeits by abandonment")


if __name__ == "__main__":
    main()
